wicketd/src/installinator_progress.rs

```diff
@@ -0,0 +1,277 @@
+        ),
+        running_updates: running_updates.clone(),
+    };
+    let ipr_update_tracker = IprUpdateTracker {
+        log: log.new(
+            slog::o!("component" => "installinator progress reports, update tracker"),
+        ),
+        running_updates,
+    };
+
+    (ipr_artifact, ipr_update_tracker)
+}
+
+/// The artifact server's interface to the installinator tracker.
+#[derive(Debug)]
+#[must_use]
+pub(crate) struct IprArtifactServer {
+    log: slog::Logger,
+    running_updates: Arc<Mutex<HashMap<Uuid, RunningUpdate>>>,
+}
+
+impl IprArtifactServer {
+    pub(crate) async fn report_progress(
+        &self,
+        update_id: Uuid,
+        report: ProgressReport,
+    ) -> ProgressReportStatus {
+        let mut running_updates = self.running_updates.lock().await;
+        if let Some(update) = running_updates.get_mut(&update_id) {
+            slog::debug!(
+                self.log,
+                "progress report seen ({} completion events, {} progress events)",
+                report.completion_events.len(),
+                report.progress_events.len();
+                "update_id" => %update_id
+            );
+            match update.take() {
+                RunningUpdate::Initial(start_sender) => {
+                    slog::debug!(
+                        self.log,
+                        "first report seen for this update ID";
+                        "update_id" => %update_id
+                    );
+                    let (sender, receiver) = mpsc::channel(16);
+                    _ = start_sender.send(receiver);
+                    *update =
+                        RunningUpdate::send_and_next_state(sender, report)
+                            .await;
+                }
+                RunningUpdate::ReportsReceived(sender) => {
+                    slog::debug!(
+                        self.log,
+                        "further report seen for this update ID";
+                        "update_id" => %update_id
+                    );
+                    *update =
+                        RunningUpdate::send_and_next_state(sender, report)
+                            .await;
+                }
+                RunningUpdate::Closed => {
+                    // The sender has been closed; ignore the report.
+                    *update = RunningUpdate::Closed;
+                }
+                RunningUpdate::Invalid => {
+                    unreachable!("invalid state")
+                }
+            }
+
+            ProgressReportStatus::Processed
+        } else {
+            slog::debug!(self.log, "update ID unrecognized"; "update_id" => %update_id);
+            ProgressReportStatus::UnrecognizedUpdateId
+        }
+    }
+}
+
+/// The update tracker's interface to the progress store.
+#[derive(Debug)]
+#[must_use]
+pub(crate) struct IprUpdateTracker {
+    log: slog::Logger,
+    running_updates: Arc<Mutex<HashMap<Uuid, RunningUpdate>>>,
```

Comment:
> Do we ever remove entries from this map?

No, not with the current design -- we move to the `RunningUpdate::Closed` status. Is there a reason removing entries would be beneficial? I figured having the idempotency is valuable.

Reply:
An insert-but-never-remove map in a long-running service leaves us growing in memory indefinitely (albeit extremely slowly in this case). Maybe just a comment that we should eventually figure out a path to garbage collect completed updates out of this map somehow? I'm not worried about it in the immediate term since updating will almost always also involve updating the scrimlets, which will result in a wicketd restart.

Reply:
Added a comment.

Reply:
Thanks!

Reply:
I think there are probably other cases like crashed installs where we'd need to clean up entries anyway so if this ever becomes an issue (really doubt it) I think we'd just want a time-based garbage collector I think.